File: LinkedList-Stack-Queue/Q145.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        # Recursive
        res = []
        self.recur_helper(root, res)
        return res

    def recur_helper(self, root, res):
        if not root:
            return

        self.recur_helper(root.left, res)
        self.recur_helper(root.right, res)
        res.append(root.val)

    def postorderTraversal(self, root: TreeNode) -> List[int]:
        # Iterative
        stack, res = [(root, False)], []
        while stack:
            node, visited = stack.pop()
            if not node:
                continue
            if visited:
                res.append(node.val)
            else:
                stack.append((node, True))
                stack.append((node.right, False))
                stack.append((node.left, False))
        return res

File: Tree/Q145.js
/**
 * Definition for a binary tree node.
 */
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val);
    this.left = (left === undefined ? null : left);
    this.right = (right === undefined ? null : right);
}


/**
 * https://leetcode.com/problems/binary-tree-postorder-traversal/
 * @param {TreeNode} root
 * @return {number[]}
 */
var postorderTraversal1 = function (root) {
    if (!root) return [];
    return [...postorderTraversal(root.left), ...postorderTraversal(root.right), root.val];
};


/**
 * https://leetcode.com/problems/binary-tree-postorder-traversal/
 * @param {TreeNode} root
 * @return {number[]}
 */
var postorderTraversal = function (root) {
    const stack = [], res = [];
    while (root || stack.length > 0) {
        while (root) {
            stack.push(root.val);
            if (root.right) stack.push(root.right);
            root = root.left;
        }

        const cur = stack.pop();
        if (cur instanceof TreeNode) root = cur;
        else res.push(cur);
    }
    return res;
};
